### PR TITLE
Bazel invoker

### LINKED ISSUE #81: Bazel Invoker
# Bazel Invoker

Bazel already supports a lot of the functionality that we need to interpret a file system. This task will be to create a wrapper around some of that functionality.

Good references for bazel cli docs (for bazel query command at least)
- [The Bazel Query Reference](https://docs.bazel.build/versions/master/query.html)
- [Bazel Query How-To](https://docs.bazel.build/versions/master/query-how-to.html)

## Look into this first

- [x] Instead of directly calling the bazel client, see if there is a way to communicate directly with the bazel server.
- [x] Probably some sort of protocol buffer.
- [x] The response would come in a more structured format.

## Create a bazel package

- [x] Create a `server.bazel` package.
- [x] Create a `server.bazel.Bazel` class which will invoke bazel commands command. Place all bazel command wrappers in here.
- [x] The structure of this package may somewhat resemble the `server.buildifier` package
- [ ] Rather than using primitives for the return type as well as parameters, create domain objects. E.g. instead of `public List<String> findRules(String ruleName, String regex, ...)`, do something like `public BazelRuleQueryOutput findRules(BazelRuleQueryInput input)`. This is just an example, your API will look different. But we want the inputs an outputs to easily extendable, so they are wrapped in a class.

## Features needed

- [ ] Find the packages that exist beneath foo. 

```
bazel query 'foo/...' --output package
```

- [x] What rules are defined in the foo package? 

`bazel query 'kind(rule, foo:*)' --output label_kind`

- [ ] What package contains file path/to/file/bar.java?

```
bazel query path/to/file/bar.java --output=package
```

- [ ] What rule target(s) contain file path/to/file/bar.java as a source?

```
fullname=$(bazel query path/to/file/bar.java)
bazel query "attr('srcs', $fullname, ${fullname//:*/}:*)"
```

- [ ] If there are other commands that you think will be useful, feel free to add those as well. A big list of common commands comes from [this page](https://docs.bazel.build/versions/master/query-how-to.html)

## QA

- [ ] Write test cases for everything you feel needs testing.

---

## Other Considerations

- We need docs to make the community happy. Provide documentation where it helps, especially on the public APIs but also on the internal APIs if it's useful.
- Try to keep as much of the implementation private as possible.
- The [legacy code](https://github.com/bazelbuild/vscode-bazel/tree/master/src/bazel) had a typescript implementation of this. Could be a good reference.

### PR DESCRIPTION
**Pull request by NovaMachina**
Adds an abstract dispatcher and the basics for any Bazel commands that we will need to create.

### CODE REVIEW

File: server/src/main/java/server/bazel/cli/Bazel.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+/**
+ * A wrapper around the Bazel Server Commands. This allows callers to invoke Bazel Server commands
+ */
+public final class Bazel {
+    private static final Logger logger = LogManager.getLogger(Bazel.class);
+    private static final CommandDispatcher dispatcher = CommandDispatcher.create("bazel-command-dispatcher");
+
+    /**
+    * Creates an instance of a Bazel.
```
Comment (josiahsrc) [inline server/src/main/java/server/bazel/cli/Bazel.java:17]: lol

File: server/src/main/java/server/bazel/cli/Bazel.java [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    */
+    private Bazel() {
+    }
+
+    public static void getBuildTargets() {
```
Comment (josiahsrc) [inline server/src/main/java/server/bazel/cli/Bazel.java:22]: Static functions may make this class hard to use in unit tests, makes it so we can't use Mockito (to the best of my knowledge). Maybe we ought to use instance methods so we can do:

```
public class SomeClassThatDependsOnBazel {
  public SomeClassThatDependsOnBazel(Bazel bzl) { ... }

  public void something()
  {
     bzl.getBuildTargets()
  }
}

...

void testSomething() {
    Bazel mockedBazel = Mockito.mock(Bazel.class);
    SomeClassThatDependsOnBazel something = new SomeClassThatDependsOnBazel(mockedBazel);
    ...
}
```

But I'm not a java buf, so someone correct me if there's a different way to do this

**Review by josiahsrc — APPROVED** [review]
Left one comment on static method usage, but other than that I approve. Nice work guys! The dispatcher is 💯

File: server/src/main/java/server/bazel/cli/Bazel.java (hunk 2 above)
Comment (NovaMachina) [reply server/src/main/java/server/bazel/cli/Bazel.java:22]: If we need to mock a static method we can use PowerMock on top of Mockito which allows the mocking of a static method.

Comment (josiahsrc) [reply server/src/main/java/server/bazel/cli/Bazel.java:22]: Perfect

**Review by jasonxris — APPROVED** [review]

**Pull request merged**